src/test_driver/happy/test-templates/ChipInetMulticast.py

```diff
@@ -0,0 +1,383 @@
+
+        if self.tap:
+            cmd += " --tap-device " + self.tap
+
+            # If present, generate and accumulate the LwIP hosted OS
+            # network local IPv4 address.
+
+            if self.ipversion == "4" and attributes.has_key('tap-ipv4-local-addr'):
+                cmd += " --local-addr " + attributes['tap-ipv4-local-addr']
+
+        # Generate and accumulate, if present, the bound network
+        # interface command line option.
+
+        if not self.interface == None:
+            cmd += " --interface " + self.interface
+
+        # Accumulate any extra command line options specified.
+
+        cmd += extra
+
+        self.logger.debug(
+            "[localhost] ChipInetMulticast: Will start process on node '%s' with tag '%s' and command '%s'" % (node, tag, cmd))
+
+        self.start_chip_process(
+            node, cmd, tag, sync_on_output=self.ready_to_service_events_str)
+
+    # Start a receiver test process on the specified node with the
+    # provided attributes and tag identifier.
+
+    def __start_receiver(self, node, attributes, identifier):
+        receiver = {}
+        tag = "INET-MCAST-RX-%u" % (identifier)
+
+        extra_cmd = " --listen"
+
+        self.__start_node(node, attributes, extra_cmd, tag)
+
+        receiver['node'] = node
+        receiver['tag'] = tag
+
+        self.receivers.append(receiver)
+
+    # Start all receiver test processes.
+
+    def __start_receivers(self):
+        receiver = {}
+        identifier = 0
+        for receiver in self.configuration['receivers']:
+            self.__start_receiver(
+                receiver, self.configuration['receivers'][receiver], identifier)
+            identifier += 1
+
+    # Start the sender test process.
+
+    def __start_sender(self):
+        node = list(self.configuration['sender'])[0]
+        attributes = self.configuration['sender'][node]
+        tag = "INET-MCAST-TX-0"
+
+        extra_cmd = ""
+
+        if not self.interval == None:
+            extra_cmd += " --interval " + str(self.interval)
+
+        extra_cmd += " --no-loopback"
+
+        self.__start_node(node, attributes, extra_cmd, tag)
+
+        self.sender['node'] = node
+        self.sender['tag'] = tag
+
+    # Block and wait for the sender test process.
+
+    def __wait_for_sender(self):
+        node = self.sender['node']
+        tag = self.sender['tag']
+
+        self.logger.debug(
+            "[localhost] ChipInetMulticast: Will wait for sender on node %s with tag %s..." % (node, tag))
+
+        self.wait_for_test_to_end(node, tag)
+
+    # Stop the specified receiver test process.
+
+    def __stop_receiver(self, receiver):
+        node = receiver['node']
+        tag = receiver['tag']
+
+        self.logger.debug(
+            "[localhost] ChipInetMulticast: Will stop receiver on node %s with tag %s..." % (node, tag))
+
+        self.stop_chip_process(node, tag)
+
+    # Stop all receiver test processes.
+    def __stop_receivers(self):
```

Comment:
👾 

Just a general comment on consistency here that at times there is a gap between the comment and function definition and some times there is not

Reply:
Should use docstrings instead. See
https://www.python.org/dev/peps/pep-0257/